invokeai/app/services/config.py

```diff
@@ -0,0 +1,290 @@
+
+  export INVOKEAI_txt2img_steps=30
+
+Order of precedence (from highest):
+   1) initialization options
+   2) command line options
+   3) environment variable options
+   4) config file options
+   5) pydantic defaults
+
+Typical usage:
+
+ from invokeai.app.services.config import InvokeAIAppConfig
+ from invokeai.invocations.generate import TextToImageInvocation
+
+ # get global configuration and print its nsfw_checker value
+ conf = InvokeAIAppConfig()
+ print(conf.nsfw_checker)
+
+ # get the text2image invocation and print its step value
+ text2image = TextToImageInvocation()
+ print(text2image.steps)
+
+'''
+import argparse
+import os
+import sys
+from argparse import ArgumentParser
+from omegaconf import OmegaConf, DictConfig
+from pathlib import Path
+from pydantic import BaseSettings, Field
+from typing import Any, ClassVar, List, Literal, Union, get_origin, get_type_hints, get_args
+
+INIT_FILE = Path('invokeai.yaml')
+
+class InvokeAISettings(BaseSettings):
+    '''
+    Runtime configuration settings in which default values are
+    read from an omegaconf .yaml file.
+    '''
+    initconf             : ClassVar[DictConfig] = None
+
+    def parse_args(self, argv: list=sys.argv[1:]):
+        parser = self.get_parser()
+        opt = parser.parse_args(argv)
+        for name in self.__fields__:
+            if name not in self._excluded():
+                setattr(self, name, getattr(opt,name))
+
+    @classmethod
+    def add_parser_arguments(cls, parser):
+        env_prefix = cls.Config.env_prefix
+        default_settings_stanza = get_args(get_type_hints(cls)['type'])[0]
+        initconf = cls.initconf.get(default_settings_stanza) if cls.initconf and default_settings_stanza in cls.initconf else None
+
+        fields = cls.__fields__
+        for name, field in fields.items():
+            if name not in cls._excluded():
+                env_name = env_prefix+f'{cls.cmd_name()}_{name}'
+                if initconf and name in initconf:
+                    field.default = initconf.get(name)
+                if env_name in os.environ:
+                    field.default = os.environ[env_name]
+                cls.add_field_argument(parser, name, field)
+
+
+    @classmethod
+    def cmd_name(self, command_field: str='type')->str:
+        hints = get_type_hints(self)
+        return get_args(hints[command_field])[0]
+
+    @classmethod
+    def get_parser(cls)->ArgumentParser:
+        parser = ArgumentParser(
+            prog=cls.cmd_name(),
+            description=cls.__doc__,
+        )
+        cls.add_parser_arguments(parser)
+        return parser
+
+    @classmethod
+    def add_subparser(cls, parser: argparse.ArgumentParser):
+        parser.add_parser(cls.cmd_name(), help=cls.__doc__)
+
+    @classmethod
+    def _excluded(self)->List[str]:
+        return ['type','initconf']
+    
+    class Config:
+        env_file_encoding = 'utf-8'
+        arbitrary_types_allowed = True
+        env_prefix = 'INVOKEAI_'
```

Comment:
I have not dug into this yet and unfortunately might not be able to until a week from now or more, but expected Pydantic environment parsing seems to not be used. `INVOKEAI_ROOT` is not seen, and `root` is still being set through `Globals`. I had `Globals` commented for troubleshooting (different comment), and this happened:

```
$ export INVOKEAI_ROOT=~/Data/invokeai-testing/pr3221
$ invokeai-node-web
...
>> InvokeAI, version 3.0.0+a0
>> InvokeAI runtime directory is "/home/clipclop/invokeai"

Reply:
I found and fixed this bug, which was only affecting the INVOKEAI_ROOT environment variable.